internal/entities/service/service.go

```diff
@@ -0,0 +1,332 @@
+		propSvc:         propSvc,
+		providerManager: providerManager,
+	}
+}
+
+//nolint:gocyclo // Let's refactor this later
+func (s *entityService) ListEntities(
+	ctx context.Context,
+	projectID uuid.UUID,
+	providerID uuid.UUID,
+	entityType pb.Entity,
+	cursor string,
+	limit int64,
+) ([]*pb.EntityInstance, string, error) {
+	l := zerolog.Ctx(ctx).With().
+		Str("projectID", projectID.String()).
+		Str("providerID", providerID.String()).
+		Str("entityType", entityType.String()).
+		Logger()
+
+	// Convert pb.Entity to db.Entities
+	dbEntityType, err := entities.EntityTypeToDBType(entityType)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to convert entity type: %w", err)
+	}
+
+	// Parse cursor if provided
+	if cursor != "" {
+		_, err := uuid.Parse(cursor)
+		if err != nil {
+			return nil, "", util.UserVisibleError(codes.InvalidArgument, "invalid cursor format")
+		}
+		// TODO: Use lastEntityID for pagination
+	}
+
+	// Set up query parameters
+	queryLimit := sql.NullInt64{Valid: false, Int64: 0}
+	if limit > 0 {
+		const maxFetchLimit = 100
+		if limit > maxFetchLimit {
+			return nil, "", util.UserVisibleError(codes.InvalidArgument, "limit too high, max is %d", maxFetchLimit)
+		}
+		queryLimit = sql.NullInt64{Valid: true, Int64: limit + 1} // +1 to check if there are more results
+	}
+
+	// Get entities from database
+	tx, err := s.store.BeginTransaction()
+	if err != nil {
+		return nil, "", fmt.Errorf("error starting transaction: %w", err)
+	}
+	defer func() {
+		if err := tx.Rollback(); err != nil && !errors.Is(err, sql.ErrTxDone) {
+			l.Error().Err(err).Msg("error rolling back transaction")
+		}
+	}()
+
+	qtx := s.store.GetQuerierWithTransaction(tx)
+
+	// TODO: Implement proper pagination with cursor
+	outentities, err := qtx.GetEntitiesByType(ctx, db.GetEntitiesByTypeParams{
+		EntityType: dbEntityType,
+		ProviderID: providerID,
+		Projects:   []uuid.UUID{projectID},
+	})
+	if err != nil {
+		return nil, "", fmt.Errorf("error fetching entities: %w", err)
+	}
+
+	// Convert to EntityWithProperties and fetch properties
+	var results []*pb.EntityInstance
+	var nextCursor string
+
+	for i, entity := range outentities {
+		// Apply limit if specified
+		if queryLimit.Valid && int64(i) >= queryLimit.Int64-1 {
+			nextCursor = entity.ID.String()
+			break
+		}
+
+		ewp, err := s.propSvc.EntityWithPropertiesByID(ctx, entity.ID,
+			propService.CallBuilder().WithStoreOrTransaction(qtx))
+		if err != nil {
+			return nil, "", fmt.Errorf("error fetching properties for entity: %w", err)
+		}
+
+		if err := s.propSvc.RetrieveAllPropertiesForEntity(ctx, ewp, s.providerManager,
+			propService.ReadBuilder().WithStoreOrTransaction(qtx).TolerateStaleData()); err != nil {
+			return nil, "", fmt.Errorf("error fetching properties for entity: %w", err)
+		}
+
+		// Convert to protobuf
+		pbEntity := entityInstanceToProto(ewp)
+
+		results = append(results, pbEntity)
+	}
+
+	// Commit transaction
+	if err := tx.Commit(); err != nil {
```

Comment:
Why do we need to commit the transaction?  There shouldn't be any writes here, correct?

Reply:
(this is still open)

Reply:
Giving this more thought, I think it's actually a good idea to leave this and explicitly COMMIT. Even if nothing is being written, this service _might_ be used in such a way that this is called first and then a write operation, which would have an unintended effect of having an unclosed transaction. But maybe I'm overthinking this, what do you think? Is this a valid concern?

Reply:
I was wondering why not simply rollback